File: cpp/problem027.cpp
// quadratic primes

#include <vector>
#include <iostream>

using namespace std;

vector<bool> prime_sieve(int n) {
    vector<bool> isprime(n+1, true);
    isprime[0] = isprime[1] = false;

    int div = 0;
    do {
        while (!isprime[++div]);
        for (int i = 2 * div; i < isprime.size(); i += div) {
            isprime[i] = false;
        }
    } while (div * div <= isprime.size());

    return isprime;
}


int main()
{
    int a_min = -1000, a_max = 1000;
    int b_min = -1000, b_max = 1000;
    int prime_max = 100000;

    auto _is_prime = prime_sieve(prime_max);
    auto is_prime = [&_is_prime](int n) -> bool { return n > 0 && _is_prime[n]; };

    int L = 0, A = 0, B = 0;
    for (int b=b_min; b <= b_max; ++b) {
        if (!is_prime(b)) {
            continue;
        }

        for (int a = a_min; a <= a_max; ++a) {
            int n =1;
            while (is_prime(n*n + a*n + b)) {
                n++;
            }

            if (n > L) {
                L = n;
                A = a;
                B = b;
            }
        }
    }

    cout << A * B << endl;

    return 0;
}

File: python/problem027.py
# quadratic primes
#
# n*n + a*n + b, find a,b which produces the longest sequence of prime numbers for 
# the consecutive values of n = 0, 1, ...
#
# b has to be a prime number, due to the n = 0 term.
#

MAX = 100000
A_RANGE = -1000, 1001
B_RANGE = -1000, 1001


def next(v):
    n = 0
    while n < len(v):
        if v[n]: yield n
        n += 1


def prime_sieve(n):
    is_prime = [True] * (n+1)
    is_prime[0] = is_prime[1] = False
    
    for p in next(is_prime):
        for x in range(p+p, n+1, p):
            is_prime[x] = False

    return is_prime


_is_prime = prime_sieve(MAX + 1)
def is_prime(n):
    if n < 0: return False
    return _is_prime[n]


L, A, B = 0, 0, 0
for b in range(*B_RANGE):
    if not is_prime(b):
        continue

    for a in range(*A_RANGE):
        n = 1
        while is_prime(n*n + n*a + b):
            n += 1

        if n > L:
            L, A, B = n, a, b


print(A*B)

